File: server/internal/lint/sizebudget_test.go
package lint

import (
	"encoding/json"
	"io/fs"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"testing"
)

// budget is the committed ratchet in sizebudget.json.
type budget struct {
	MaxNewFileLines int               `json:"maxNewFileLines"`
	Files           map[string]int    `json:"files"`
	CommentPct      map[string]int    `json:"commentPct"`
	Exceptions      map[string]string `json:"exceptions"`
}

type counts struct{ Total, Comment int }

// skipDirs are never scanned: vendor is third-party, data is runtime output.
var skipDirs = map[string]bool{"vendor": true, "data": true, "node_modules": true, ".git": true}

func here() string {
	_, file, _, _ := runtime.Caller(0)
	return filepath.Dir(file)
}

// load reads the budget and scans the tree it governs. Paths are relative to the
// repo root (nearest .git above this file), so they read the same anywhere.
func load(t *testing.T) (budget, map[string]counts) {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(here(), "sizebudget.json"))
	if err != nil {
		t.Fatalf("read budget: %v", err)
	}
	var b budget
	if err := json.Unmarshal(raw, &b); err != nil {
		t.Fatalf("parse budget: %v", err)
	}
	if b.MaxNewFileLines <= 0 {
		t.Fatal("budget: maxNewFileLines must be positive")
	}
	return b, scanTree(t, repoRoot(t))
}

func repoRoot(t *testing.T) string {
	t.Helper()
	for dir := here(); ; {
		if _, err := os.Stat(filepath.Join(dir, ".git")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			t.Fatalf("no repo root (.git) above %s", here())
		}
		dir = parent
	}
}

// scanTree returns every in-scope .go file under server/, keyed by repo-relative path.
func scanTree(t *testing.T, root string) map[string]counts {
	t.Helper()
	out := map[string]counts{}
	base := filepath.Join(root, "server")
	err := filepath.WalkDir(base, func(path string, d fs.DirEntry, err error) error {
		switch {
		case err != nil:
			return err
		case d.IsDir() && skipDirs[d.Name()]:
			return fs.SkipDir
		case d.IsDir(), filepath.Ext(path) != ".go":
			return nil
		}
		rel, err := filepath.Rel(root, path)
		if err != nil {
			return err
		}
		out[filepath.ToSlash(rel)] = countFile(t, path)
		return nil
	})
	if err != nil {
		t.Fatalf("walk %s: %v", base, err)
	}
	if len(out) == 0 {
		t.Fatalf("no .go files found under %s", base)
	}
	return out
}

func countFile(t *testing.T, path string) counts {
	t.Helper()
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	text := strings.TrimSuffix(string(raw), "\n")
	if text == "" {
		return counts{}
	}
	c, st := counts{}, scanState{}
	for _, line := range strings.Split(text, "\n") {
		var isComment bool
		isComment, st = scanLine(line, st)
		c.Total++
		if isComment {
			c.Comment++
		}
	}
	return c
}

// scanState carries string/comment context across line breaks.
type scanState struct{ inBlock, inRaw bool }

// scanLine reports whether the line opens with a comment (or sits inside a block comment) and returns the
// state for the next line. It steps over string, rune and raw-string literals.
func scanLine(line string, st scanState) (bool, scanState) {
	comment := st.inBlock
	i := 0
	if !st.inBlock && !st.inRaw {
		for i < len(line) && (line[i] == ' ' || line[i] == '\t') {
			i++
		}
		if strings.HasPrefix(line[i:], "//") {
			return true, st
		}
		comment = strings.HasPrefix(line[i:], "/*")
	}
	for i < len(line) {
		switch {
		case st.inBlock, st.inRaw:
			term := "*/"
			if st.inRaw {
				term = "`"
			}
			end := strings.Index(line[i:], term)
			if end < 0 {
				return comment, st
			}
			st.inBlock, st.inRaw, i = false, false, i+end+len(term)
		case strings.HasPrefix(line[i:], "//"):
			return comment, st
		case strings.HasPrefix(line[i:], "/*"):
			st.inBlock, i = true, i+2
		case line[i] == '`':
			st.inRaw, i = true, i+1
		case line[i] == '"' || line[i] == '\'':
			i = skipQuoted(line, i)
		default:
			i++
		}
	}
	return comment, st
}

// skipQuoted returns the index past the literal opened at i (len(line) if unterminated).
func skipQuoted(line string, i int) int {
	for j := i + 1; j < len(line); j++ {
		if line[j] == '\\' {
			j++
		} else if line[j] == line[i] {
			return j + 1
		}
	}
	return len(line)
}

// Unlisted files stay under the cap, listed files never grow; "exceptions"
// (generated bundles, byte-pinned twins) sit outside both.
func TestFileSizeBudget(t *testing.T) {
	b, files := load(t)
	for _, path := range sortedKeys(files) {
		if _, skipped := b.Exceptions[path]; skipped {
			continue
		}
		total := files[path].Total
		recorded, listed := b.Files[path]
		switch {
		case !listed && total > b.MaxNewFileLines:
			t.Errorf("%s is %d lines, over the %d-line cap for unlisted files; split it or add it to sizebudget.json", path, total, b.MaxNewFileLines)
		case listed && total > recorded:
			t.Errorf("%s grew to %d lines, budget is %d; shrink it instead of raising the number", path, total, recorded)
		case listed && total <= recorded*9/10:
			t.Logf("%s shrank to %d lines (budget %d) — ratchet sizebudget.json down", path, total, recorded)
		}
	}
}

// No directory's comment share may rise above its recorded percent.
func TestCommentShareBudget(t *testing.T) {
	b, files := load(t)
	dirs := map[string]counts{}
	for path, c := range files {
		dir := filepath.ToSlash(filepath.Dir(path))
		agg := dirs[dir]
		agg.Total, agg.Comment = agg.Total+c.Total, agg.Comment+c.Comment
		dirs[dir] = agg
	}
	for _, dir := range sortedKeys(b.CommentPct) {
		if got := pct(dirs[dir]); got > b.CommentPct[dir] {
			t.Errorf("%s comment share is %d%%, budget is %d%%", dir, got, b.CommentPct[dir])
		}
	}
}

func pct(c counts) int {
	if c.Total == 0 {
		return 0
	}
	return c.Comment * 100 / c.Total
}

func sortedKeys[V any](m map[string]V) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

File: e2e/tests/sizeBudget.test.js
// Size + comment ratchet for e2e/ (.js under helpers/, tests/ and the config), the twin of
// browser/tests/sizeBudget.test.js. Nothing new may land over maxNewFileLines, nothing listed
// may grow, and no directory's comment share may rise. Budget lives in sizeBudget.json.
import test from 'node:test';
import assert from 'node:assert';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import config from '../playwright.config.js';

const HERE = path.dirname(fileURLToPath(import.meta.url));

// Walk up to the repo root so every recorded path reads the same from anywhere.
const repoRoot = () => {
  let dir = HERE;
  while (!fs.existsSync(path.join(dir, '.git'))) {
    const up = path.dirname(dir);
    assert.notStrictEqual(up, dir, 'repo root not found above ' + HERE);
    dir = up;
  }
  return dir;
};

const ROOT = repoRoot();
const SCAN = ['e2e'];
const SKIP_DIRS = new Set(['node_modules', 'test-results', 'playwright-report', 'blob-report', '.playwright']);
const budget = JSON.parse(fs.readFileSync(path.join(HERE, 'sizeBudget.json'), 'utf8'));
const MAX = budget.maxNewFileLines;

const walk = (rel, out = []) => {
  const entries = fs.readdirSync(path.join(ROOT, rel), { withFileTypes: true });
  for (const e of entries.sort((a, b) => (a.name < b.name ? -1 : 1))) {
    const child = `${rel}/${e.name}`;
    if (e.isDirectory()) { if (!SKIP_DIRS.has(e.name)) walk(child, out); }
    else if (e.name.endsWith('.js')) out.push(child);
  }
  return out;
};

// A line counts as a comment when it opens with // or /*, or sits inside a block comment; the scanner tracks
// strings, so a // or /* inside a literal opens nothing (' and " cannot span lines, ` can).
const measure = (rel) => {
  const lines = fs.readFileSync(path.join(ROOT, rel), 'utf8').split('\n');
  if (lines.length && lines[lines.length - 1] === '') lines.pop();
  let block = false, tmpl = false, comments = 0;
  for (const line of lines) {
    const trimmed = line.trimStart();
    if (block || (!tmpl && (trimmed.startsWith('//') || trimmed.startsWith('/*')))) comments++;
    let quote = tmpl ? '`' : '';
    for (let i = 0; i < line.length; i++) {
      const c = line[i], n = line[i + 1];
      if (block) { if (c === '*' && n === '/') { block = false; i++; } continue; }
      if (quote) { if (c === '\\') i++; else if (c === quote) quote = ''; continue; }
      if (c === '/' && n === '/') break;
      if (c === '/' && n === '*') { block = true; i++; continue; }
      if (c === '"' || c === "'" || c === '`') quote = c;
    }
    tmpl = quote === '`';
  }
  return { total: lines.length, comments };
};

const files = SCAN.flatMap((rel) => walk(rel));
const sizes = new Map(files.map((rel) => [rel, measure(rel)]));

test('size budget: no new file over the cap, no listed file grew', () => {
  const appeared = [], grew = [], shrank = [];
  for (const [rel, { total }] of sizes) {
    if (rel in budget.exceptions) continue;
    const cap = budget.files[rel];
    if (cap === undefined) {
      if (total > MAX) appeared.push(`${rel} is ${total} lines (cap ${MAX})`);
      continue;
    }
    if (total > cap) grew.push(`${rel} grew to ${total} lines (budget ${cap})`);
    else if (total < cap * 0.9) shrank.push(`${rel}: ${cap} → ${total}`);
  }
  if (shrank.length) console.log(`  note: ratchet these down in sizeBudget.json —\n    ${shrank.join('\n    ')}`);
  assert.deepStrictEqual(appeared, [], 'new oversized files must be split, or recorded in sizeBudget.json');
  assert.deepStrictEqual(grew, [], 'listed files may shrink, never grow');
});

test('size budget: every recorded path still exists and is in scope', () => {
  const stale = [...Object.keys(budget.files), ...Object.keys(budget.exceptions)].filter((rel) => !sizes.has(rel));
  assert.deepStrictEqual(stale, [], 'drop these from sizeBudget.json');
  for (const [rel, why] of Object.entries(budget.exceptions)) assert.ok(why, `exception ${rel} needs a reason`);
});

const stemsPerDir = () => {
  const dirs = new Map();
  for (const rel of sizes.keys()) {
    const dir = path.posix.dirname(rel);
    if (!dirs.has(dir)) dirs.set(dir, new Set());
    dirs.get(dir).add(path.posix.basename(rel).slice(0, -3));
  }
  return dirs;
};

test('folder fan-out: no directory over the cap, no listed directory grew', () => {
  const cap = budget.maxFilesPerDir;
  const appeared = [], grew = [], shrank = [];
  for (const [dir, stems] of stemsPerDir()) {
    const n = stems.size;
    const allowed = budget.dirs[dir];
    if (allowed === undefined) {
      if (n > cap) appeared.push(`${dir} holds ${n} modules (cap ${cap}) — split it by feature`);
      continue;
    }
    if (n > allowed) grew.push(`${dir} grew to ${n} modules (budget ${allowed})`);
    else if (n < allowed) shrank.push(`'${dir}': ${n}`);
  }
  const gone = Object.keys(budget.dirs).filter((d) => !stemsPerDir().has(d));
  if (shrank.length) console.log(`  note: lower these in sizeBudget.json —\n    ${shrank.join('\n    ')}`);
  assert.deepStrictEqual(appeared, [], 'a new folder over the cap must be split, never recorded');
  assert.deepStrictEqual(grew, [], 'listed folders may shrink, never grow');
  assert.deepStrictEqual(gone, [], 'drop these from sizeBudget.json — the folder is gone');
});

test('comment share per directory did not rise', () => {
  const dirs = new Map();
  for (const [rel, { total, comments }] of sizes) {
    const dir = path.posix.dirname(rel);
    const acc = dirs.get(dir) || { total: 0, comments: 0 };
    acc.total += total; acc.comments += comments;
    dirs.set(dir, acc);
  }
  const over = [];
  for (const [dir, cap] of Object.entries(budget.commentPct)) {
    const acc = dirs.get(dir);
    assert.ok(acc, `${dir} is in sizeBudget.json but holds no .js files`);
    const pct = Math.floor(acc.comments * 100 / acc.total);
    if (pct > cap) over.push(`${dir} is ${pct}% comments (budget ${cap}%)`);
  }
  assert.deepStrictEqual(over, [], 'trim the prose, or split the code out of these dirs');
});

// Test-count floor: Playwright can only be asked what it runs by running it, so this counts
// `test(` declarations instead — it catches a deleted spec, not an unexecuted one.
const TEST_FLOOR = 260;
const DECL = /^\s*test(?:\.\w+)*\s*\(/gm;

test('test-count floor: the spec tree still declares its cases', () => {
  let count = 0;
  for (const rel of sizes.keys()) {
    if (!rel.endsWith('.spec.js')) continue;
    count += (fs.readFileSync(path.join(ROOT, rel), 'utf8').match(DECL) || []).length;
  }
  assert.ok(count >= TEST_FLOOR, `e2e suite collapsed to ${count} declared tests, floor is ${TEST_FLOOR}`);
});

// The runner's own blind spot: every project sets its own testMatch, so a spec in a directory
// no project names is collected by nothing and reports neither pass nor skip.
test('every spec is claimed by a playwright project', () => {
  const orphans = [...sizes.keys()].filter((rel) => rel.endsWith('.spec.js'))
    .filter((rel) => !config.projects.some((p) => p.testMatch.test(rel.slice('e2e/'.length))));
  assert.deepStrictEqual(orphans, [], 'add a project testMatch in playwright.config.js, or move the spec');
});
